ndk/src/midi.rs

```diff
@@ -0,0 +1,297 @@
+//! Bindings for [`AMidiDevice`], [`AMidiInputPort`], and [`AMidiOutputPort`]
+//!
+//! See [the NDK guide](https://developer.android.com/ndk/guides/audio/midi) for
+//! design and usage instructions, and [the NDK reference](https://developer.android.com/ndk/reference/group/midi)
+//! for an API overview.
+//!
+//! [`AMidiDevice`]: https://developer.android.com/ndk/reference/group/midi#amididevice
+//! [`AMidiInputPort`]: https://developer.android.com/ndk/reference/group/midi#amidiinputport
+//! [`AMidiOutputPort`]: https://developer.android.com/ndk/reference/group/midi#amidioutputport
+#![cfg(feature = "midi")]
+
+pub use super::media::Result;
+use super::media::{construct, NdkMediaError};
+
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+use std::convert::TryFrom;
+use std::fmt;
+use std::marker::PhantomData;
+use std::os::raw::{c_int, c_uint};
+use std::ptr::NonNull;
+
+// There is no mention about thread-safety in the NDK reference, but the official Android C++ MIDI
+// sample stores `AMidiDevice *` and `AMidi{Input,Output}Port *` in global variables and accesses the
+// ports from separate threads.
+// See https://github.com/android/ndk-samples/blob/7f6936ea044ee29c36b5c3ebd62bb3a64e1e6014/native-midi/app/src/main/cpp/AppMidiManager.cpp
+unsafe impl Send for MidiDevice {}
+unsafe impl<'a> Send for MidiInputPort<'a> {}
+unsafe impl<'a> Send for MidiOutputPort<'a> {}
```

Comment:
Can they be accessed concurrently (i.e., `Sync`), or are these only safe to be used on different threads as long as there aren't two or more threads poking it at the same time (`Send`)?

Reply:
`MidiIntputPort` is definitely not `Sync`: it puts `write` in the loop, so there's a possibility of interleaved writing.
![image](https://user-images.githubusercontent.com/17005454/194055808-844c8be7-38f1-40cd-9a30-e7809f5dada6.png)

`MidiOutputPort` is definitely `Sync`:`MidiReceiver`, which is called by `AMidiOutputPort_receive` implements synchronization logic. `MidiOutputPort::receive` will immediately return an `Err` if a simultaneous read call was invoked by another thread.
![image](https://user-images.githubusercontent.com/17005454/194058709-60a8825a-f714-4950-887b-404cefcf84c5.png)

`MidiDevice` seems to be `Sync`: as `MidiDeviceServer`, which is called by `MidiDevice`, uses `synchronized` in every method except for `getDeviceInfo()` and `setDeviceInfo()`.
![image](https://user-images.githubusercontent.com/17005454/194057510-0ed59a6b-9072-4381-9578-793921d1e80a.png)

Reply:
That's an interesting design choice and discrepancy between input and output ports. We can also enforce the write with a `&mut self` borrow but I think keeping them as `Send` without `Clone`/`Copy` is also fine.

Reply:
I re-read the source code of `MidiDevice` and `MidiDeviceServer` - I think we can assume that `MidiDevice` is `Sync` as well.

Reply:
Would the lack of mentions of thread safety in the NDK reference mean that there would be potential changes that make `AMidiDevice` thread-unsafe in the future version?

Reply:
Let's not preemptively add `Send`/`Sync` in that case. If it's not documented there's nothing holding the NDK folks from changing the underling implementation and comments.

We recently [requested clarification](https://github.com/android/ndk/discussions/1768) on other APIs as well, perhaps you can do the same here?

TLDR: Let's keep it `Send` for now until we have official confirmation for `Sync`.

Reply:
I found that `AMidiDevice` must be released in the thread where the JNI is attached, i.e., the app crashes when `AMidiDevice_release` is called before `AttachCurrentThread[AsDaemon]` is called from that thread. I'll remove `Send` from `MidiDevice`.

```rust
let midi_device = MidiDevice::from_java(...);
thread::spawn(move || {
  midi_device. ...
  // JNIEnv access error on AMidiDevice_release
});
```

```rust
let midi_device = MidiDevice::from_java(...);
thread::spawn(move || {
  AttachCurrentThread(vm, &env, ...);
  midi_device. ...
  // No error
});
```

Reply:
Looks like you resolved this. No clue about removing `impl Send` from the ports, the NDK makes no guarantees but it seems logical that you might want to process such data on another thread even if the JNI device can only be accessed on a VM-attached thread?